

# Index database files with **tabix**:
* [Index a BED database file](#index-a-bed-database-file)
* [Index a VCF database file](#index-a-vcf-database-file)


## Index a BED database file
The instructions provided here can be used to create any BED database files.  For simplicity, a simple and an example BED file called _target.bed_ is provided in _${QPIPELINE_HOME}/test_data/vcf_.  

Change directory to _${QPIPELINE_HOME}/test_data/vcf_.  To create a BED database file for _target.bed_, do the following
```
# set FILE so we can cut and paste and reuse all the commands below 
# for different BED database file when needed 
FILE="target.bed"

#sort the file by chromosomes, start, and end
cat $FILE  | sort -k1,1 -k2,2n -k3,3n > ${FILE}.sorted

# Zip the sorted file using bgzip ( bgzip is part of tabix )
bgzip ${FILE}.sorted


# Index the zipped file using tabix
# An index file with extension .tbi is generated by tabix
# Once done, ${FILE}.sorted.gz  can be used as the BED database for qpipeline
tabix -p bed ${FILE}.sorted.gz 
```


## Create and index a VCF database file 
The instructions provided here can be used to create any VCF database files.  For simplicity, a partial VCF file from NCBI called _common_all_20161122.vcf_ is provided in _${QPIPELINE_HOME}/test_data/vcf_ directory.  This file contained only small number of entries from chromsome 1 downloaded from ftp://ftp.ncbi.nlm.nih.gov/snp/organisms/human_9606/VCF/common_all_20161122.vcf.gz

NOTE: Both the input VCF file and the VCF database file must have the same chromosome format. 
That is, either '1',..., '22', 'X', 'Y', 'M' or 'chr1',..., 'chr22', 'chrX', 'chrY', 'chrM'.  Check to make sure your input VCF file and your VCF database file have the same format before indexing your VCF database file.

The sample input VCF file ( _${QPIPELINE_HOME}/test_data/vcf/sample.vcf_ ) uses 'chr1',...,'chr22', 'chrX', 'chrY','chrM' so we will add the prefix 'chr' to _common_all_20161122.vcf_ before indexing it.

Change directory to _${QPIPELINE_HOME}/test_data/vcf_.  To create a tabix database file for _common_all_20161122.vcf_, do the following

```
# set FILE variable so we can cut and paste and reuse all the commands below 
# for different VCF database file when needed 
FILE="common_all_20161122.vcf"

# get just the VCF header then add 'chr' to all entries
cat $FILE | grep ^# >  ${FILE}.modified.vcf
cat $FILE | grep -v ^# | awk '{ print "chr"$0 }' >> ${FILE}.modified.vcf

# Zip and index using **tabix**

bgzip  ${FILE}.modified.vcf

# Index the zipped file using tabix
# An index file with extension .tbi is generated by tabix
# Once done, ${FILE}.sorted.gz  can be used as the VCF database for qpipeline
tabix -p vcf  ${FILE}.modified.vcf.gz 
```


